puree/src/main/java/com/mercari/puree/TagPattern.java

```diff
@@ -0,0 +1,96 @@
+package com.mercari.puree;
+
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+public class TagPattern {
+
+    private static final String SEPARATOR = "\\.";
+    private static final String ALL_WILD_CARD = "**";
+    private static final String WILD_CARD = "*";
+    private static final Pattern VALID_PATTERN_ALPHANUMERIC = Pattern.compile("^[A-Za-z0-9]+");
+
+    private String pattern;
+
+    public TagPattern() {
+        this.pattern = ALL_WILD_CARD;
+    }
+
+    private TagPattern(String pattern) {
+        this.pattern = pattern;
+    }
+
+    /**
+     * Patterns should follow the following pattern:
+     * • No consecutive periods
+     * • Accepts only alphanumeric, '*', '**', and '.'
+     * @param pattern to be matched
+     * @return whether pattern is valid
+     */
+    private static boolean isValidPattern(String pattern) {
+        String[] patternElements = pattern.split(SEPARATOR);
```

Comment:
It will crash with a NPE if pattern is null. Is it expected behavior?

Reply:
Not a big deal but considering fromString is returning null if an error happened, better do the same if pattern is null for consistency.

Reply:
Added a null check